EDMarketConnector.py

```diff
@@ -18,74 +18,69 @@
-                        and window_title(window_handle) == applongname \
-                        and GetProcessHandleFromHwnd(window_handle):
-                    # If GetProcessHandleFromHwnd succeeds then the app is already running as this user
-                    if len(sys.argv) > 1 and sys.argv[1].startswith(protocolhandler.redirect):
-                        CoInitializeEx(0, COINIT_APARTMENTTHREADED | COINIT_DISABLE_OLE1DDE)
-                        # Wait for it to be responsive to avoid ShellExecute recursing
-                        ShowWindow(window_handle, SW_RESTORE)
-                        ShellExecute(0, None, sys.argv[1], None, None, SW_RESTORE)
-
-                    else:
-                        ShowWindowAsync(window_handle, SW_RESTORE)
-                        SetForegroundWindow(window_handle)
-
-                    return False
-
-                return True
-
-            return EnumWindows(enumwindowsproc, 0)
+            print('no_other_instance_running(): win32, using msvcrt')
+            # win32 doesn't have fcntl, so we have to use msvcrt
+            import msvcrt
+
+            print(f'no_other_instance_running(): journal_dir_lockfile = {journal_dir_lockfile!r}')
+
+            try:
+                msvcrt.locking(journal_dir_lockfile.fileno(), msvcrt.LK_NBLCK, 4096)
+
+            except PermissionError as e:
+                print(f"PermissionError: Couldn't lock journal directory \"{journal_dir}\", assuming another process running\n{e}")
+                return False
 
+            except OSError as e:
+                print(f"OSError: Couldn't lock journal directory \"{journal_dir}\", assuming another process running\n{e}")
+                return False
+
+            except Exception as e:
+                print(f"other Exception: Couldn't lock journal directory \"{journal_dir}\", assuming another process running\n{e}")
+                return False
+
+            journal_dir_lockfile.write(f"Path: {journal_dir}\nPID: {os_getpid()}\n")
+
+        else:
+            print('no_other_instance_running(): NOT win32, using fcntl')
+            try:
+                import fcntl
+
+            except Exception:
+                print("Not on win32 and we have no fcntl, can't use a file lock!  Allowing multiple instances!")
+
+            try:
+                fcntl.flock(journal_dir_lockfile, fcntl.LOCK_EX | fcntl.LOCK_NB)
+
+            except BlockingIOError as e:
+                print(f"BlockingIOError: Couldn't lock journal directory \"{journal_dir}\", assuming another process running\n{e}")
+                return False
+
+            except OSError as e:
+                print(f"OSError: Couldn't lock journal directory \"{journal_dir}\", assuming another process running\n{e}")
+                return False
+
+            except Exception as e:
+                print(f"other Exception: Couldn't lock journal directory \"{journal_dir}\", assuming another process running\n{e}")
+                return False
```

Comment:
```suggestion
            except (BlockingIOError, OSError, Exception) as e:
                print(f'{type(e).__name__}: Couldn't lock journal directory "{journal_dir}", assuming another process running\n{e}')
                return False
```

Reply:
This was paranoia brought on by the initial problems I had getting it working at all.  Given *every* other platform will use fcntl version, and I have no way to test it on e.g. MacOS, I wanted to leave this in so it goes from more specific to less to be sure we print *something* for when it doesn't work as expected.

Reply:
Yeah, I just thought that three different except branches that all do the same thing is silly. Its still catching all three, and still prints the exception name as it goes, just all in one rather than as three